// chimpanzee/dist/operators/types.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.number = number;
exports.bool = bool;
exports.string = string;
exports.object = object;
exports.func = func;

var _capture = require("./capture");

var _results = require("../results");

var _schemas = require("../schemas");

var _parse = require("../parse");

var _parse2 = _interopRequireDefault(_parse);

var _utils = require("./utils");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function number(params) {
  return checkType("number", params);
} /*       */
function bool(params) {
  return checkType("boolean", params);
}

function string(params) {
  return checkType("string", params);
}

function object(params) {
  return checkType("object", params);
}

function func(params) {
  return checkType("function", params);
}

function checkType(type, params = {}) {
  const meta = { type, params };

  function fn(obj, key, parents, parentKeys) {
    return function (context) {
      const result = (0, _parse2.default)((0, _capture.captureIf)(function (obj) {
        return typeof obj === type;
      }))(obj, key, parents, parentKeys)(context);
      return result instanceof _results.Skip ? new _results.Skip(`Expected ${type} but got ${typeof obj}.`, { obj, key, parents, parentKeys }, meta) : result;
    };
  }

  return new _schemas.FunctionSchema(fn, (0, _utils.getParams)(params), meta);
}
