// chimpanzee/dist/chimpanzee.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.builtins = exports.Fault = exports.Skip = exports.Empty = exports.Match = exports.wrap = exports.array = exports.optionalItem = exports.unorderedItem = exports.repeatingItem = exports.regex = exports.optional = exports.func = exports.object = exports.string = exports.bool = exports.number = exports.map = exports.exists = exports.empty = exports.deep = exports.composite = exports.take = exports.modify = exports.literal = exports.captureAndParse = exports.captureIf = exports.capture = exports.any = exports.parse = undefined;

var _parse = require("./parse");

Object.defineProperty(exports, "parse", {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_parse).default;
  }
});

var _any = require("./operators/any");

Object.defineProperty(exports, "any", {
  enumerable: true,
  get: function get() {
    return _any.any;
  }
});

var _capture = require("./operators/capture");

Object.defineProperty(exports, "capture", {
  enumerable: true,
  get: function get() {
    return _capture.capture;
  }
});
Object.defineProperty(exports, "captureIf", {
  enumerable: true,
  get: function get() {
    return _capture.captureIf;
  }
});
Object.defineProperty(exports, "captureAndParse", {
  enumerable: true,
  get: function get() {
    return _capture.captureAndParse;
  }
});
Object.defineProperty(exports, "literal", {
  enumerable: true,
  get: function get() {
    return _capture.literal;
  }
});
Object.defineProperty(exports, "modify", {
  enumerable: true,
  get: function get() {
    return _capture.modify;
  }
});
Object.defineProperty(exports, "take", {
  enumerable: true,
  get: function get() {
    return _capture.take;
  }
});

var _composite = require("./operators/composite");

Object.defineProperty(exports, "composite", {
  enumerable: true,
  get: function get() {
    return _composite.composite;
  }
});

var _deep = require("./operators/deep");

Object.defineProperty(exports, "deep", {
  enumerable: true,
  get: function get() {
    return _deep.deep;
  }
});

var _empty = require("./operators/empty");

Object.defineProperty(exports, "empty", {
  enumerable: true,
  get: function get() {
    return _empty.empty;
  }
});

var _exists = require("./operators/exists");

Object.defineProperty(exports, "exists", {
  enumerable: true,
  get: function get() {
    return _exists.exists;
  }
});

var _map = require("./operators/map");

Object.defineProperty(exports, "map", {
  enumerable: true,
  get: function get() {
    return _map.map;
  }
});

var _types = require("./operators/types");

Object.defineProperty(exports, "number", {
  enumerable: true,
  get: function get() {
    return _types.number;
  }
});
Object.defineProperty(exports, "bool", {
  enumerable: true,
  get: function get() {
    return _types.bool;
  }
});
Object.defineProperty(exports, "string", {
  enumerable: true,
  get: function get() {
    return _types.string;
  }
});
Object.defineProperty(exports, "object", {
  enumerable: true,
  get: function get() {
    return _types.object;
  }
});
Object.defineProperty(exports, "func", {
  enumerable: true,
  get: function get() {
    return _types.func;
  }
});

var _optional = require("./operators/optional");

Object.defineProperty(exports, "optional", {
  enumerable: true,
  get: function get() {
    return _optional.optional;
  }
});

var _regex = require("./operators/regex");

Object.defineProperty(exports, "regex", {
  enumerable: true,
  get: function get() {
    return _regex.regex;
  }
});

var _array = require("./operators/array");

Object.defineProperty(exports, "repeatingItem", {
  enumerable: true,
  get: function get() {
    return _array.repeatingItem;
  }
});
Object.defineProperty(exports, "unorderedItem", {
  enumerable: true,
  get: function get() {
    return _array.unorderedItem;
  }
});
Object.defineProperty(exports, "optionalItem", {
  enumerable: true,
  get: function get() {
    return _array.optionalItem;
  }
});
Object.defineProperty(exports, "array", {
  enumerable: true,
  get: function get() {
    return _array.array;
  }
});

var _wrap = require("./operators/wrap");

Object.defineProperty(exports, "wrap", {
  enumerable: true,
  get: function get() {
    return _wrap.wrap;
  }
});

var _results = require("./results");

Object.defineProperty(exports, "Match", {
  enumerable: true,
  get: function get() {
    return _results.Match;
  }
});
Object.defineProperty(exports, "Empty", {
  enumerable: true,
  get: function get() {
    return _results.Empty;
  }
});
Object.defineProperty(exports, "Skip", {
  enumerable: true,
  get: function get() {
    return _results.Skip;
  }
});
Object.defineProperty(exports, "Fault", {
  enumerable: true,
  get: function get() {
    return _results.Fault;
  }
});
exports.match = match;

var _parse2 = _interopRequireDefault(_parse);

var _builtins = require("./operators/builtins");

var builtins = _interopRequireWildcard(_builtins);

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

exports.builtins = builtins;
function match(schema, input) {
  return (0, _parse2.default)(schema)(input, "__INIT__")();
}
